### PR TITLE
PowerShellForGitHub: Improve and Standardise WhatIf/Confirm Processing

### LINKED ISSUE #225: PowerShellForGitHub: Suggested WhatIf/Confirm Processing Improvements
#### Issue Details

Since PR #213, the `ShouldProcess` (WhatIf/Confirm) processing in the module is better, but it still could be improved.
The `PowerShellForGuthub` functions are designed to be consumed by other scripts/functions,
so the norm in this case is for each function to only ever output a single `ShouldProcess` prompt.
Currently, most functions output at least two prompts, for example:

```powershell
PS> New-GitHubRepository -OrganizationName test -RepositoryName test -WhatIf
What if: Performing the operation "Output to File" on target "\\nas01\data\users\simon\documents\PowerShellForGitHub.log".
What if: Performing the operation "Invoke-WebRequest" on target "https://api.github.com/orgs/test/repos".
```

The `WhatIf` for the 'Output to File' should not be shown, and the second `WhatIf` has a non-useful operation and target. It should be: `What if: Performing the operation "Create Repository" on target "test"`.

Also, if a function is given an invalid set of parameters, for example:

```powershell
New-GitHubRepository -RepositoryName test -TeamId 1 -WhatIf
What if: Performing the operation "Output to File" on target "\\nas01\data\users\simon\documents\PowerShellForGitHub.log".
Exception: Z:\Users\Simon\Documents\GitHub\X-Guardian\PowerShellForGitHub\GitHubRepositories.ps1:144:9
Line |
 144 |          throw $message
     |          ~~~~~~~~~~~~~~
     | TeamId may only be specified when creating a repository under an organization.
```     

The `WhatIf` for the 'Output to File' should not be shown.

#### Suggested solution to the issue

- Move the `$PSCmdlet.ShouldProcess` check out of `Invoke-GHRestMethod` and add it to the calling functions with the correct operation and target (as has been done for the `ConfirmHighImpact` PR #174).
- Move the `Write-InvocationLog -Invocation $MyInvocation` line in each resource function to within the `$PSCmdlet.ShouldProcess` condition block. This will prevent the log being updated if `ShouldProcess` is $false.
- Add `-WhatIf:$false` to the `Out-File` Cmdlet call in the `Write-Log` function in the `Helpers` module. This will prevent the `Write-Log` function displaying `WhatIf`.

This would also have the additional benefit of being able to remove the `SuppressMessageAttribute` `PSShouldProcess` decorator from all the functions, as the `ShouldProcess` logic will now be happening within the function.

#### Requested Assignment

- If possible, I would like to fix this.

#### Operating System

```
OsName               : Microsoft Windows 10 Pro
OsOperatingSystemSKU : 48
OsArchitecture       : 64-bit
WindowsVersion       : 1903
WindowsBuildLabEx    : 18362.1.amd64fre.19h1_release.190318-1202
OsLanguage           : en-GB
OsMuiLanguages       : {en-GB, en-US}
```

#### PowerShell Version

```
Name                           Value
----                           -----
PSVersion                      7.0.1
PSEdition                      Core
GitCommitId                    7.0.1
OS                             Microsoft Windows 10.0.18362
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

#### Module Version

Running: 0.14.0
Installed: 0.14.0

### PR DESCRIPTION
**Pull request by X-Guardian**
#### Description

This PR improves and standardises the WhatIf/Confirm ('ShouldProcess') processing across all the functions.

##### Summary of changes

- Moved the `$PSCmdlet.ShouldProcess` check out of `Invoke-GHRestMethod` and added it to the calling functions with the correct operation and target.
- Moved the `Write-InvocationLog` line in each resource function to after the `$PSCmdlet.ShouldProcess` condition block. This will prevent the log being updated if `ShouldProcess` is `$false`.
- Added `WhatIf:$false` and `Confirm:$false` to the `Out-File` Cmdlet call in the `Write-Log` function in the Helpers module. This will prevent the `Write-Log` function displaying `WhatIf` and `Confirm` prompts.
- Removed `ShouldProcess` from non-state changing functions.
- Modified the current `ShouldProcess` conditions to use an early return.
- Modified the `Set-GitHubIssueLabel` function to remove `ConfirmImpact='High'` and set `ConfirmPreference` to 'Low' if no labels have been specified instead.
- Modified the `Update-GitHubRepository` function to remove `ConfirmImpact='High'` and set `ConfirmPreference` to 'Low' if the repo is being renamed instead.

#### Issues Fixed

- Fixes #225

#### References

N/A

#### Checklist

- [ ] You actually ran the code that you just wrote, especially if you did just "one last quick change".
- [ ] ~~Comment-based help added/updated, including examples.~~
- [x] [Static analysis](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#static-analysis) is reporting back clean.
- [x] New/changed code adheres to our [coding guidelines](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#coding-guidelines).
- [x] New/changed code continues to [support the pipeline](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#pipeline-support).
- [ ] ~~Changes to the manifest file follow the [manifest guidance](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#module-manifest).~~
- [ ] ~~Unit tests were added/updated and are all passing. See [testing guidelines](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#testing).  This includes making sure that all pipeline input variations have been covered.~~
- [ ] ~~Relevant usage examples have been added/updated in [USAGE.md](https://github.com/microsoft/PowerShellForGitHub/blob/master/USAGE.md).~~
- [ ] ~~If desired, ensure your name is added to our [Contributors list](https://github.com/microsoft/PowerShellForGitHub/blob/master/CONTRIBUTING.md#contributors)~~

### CODE REVIEW

**Comment by X-Guardian** [comment]
Ran the whole test suite for this PR, and the `GitHubRepositoryForks` test has deleted my fork of the `PowerShellForGitHub` repo, which includes all the branches for the pending PRs here including this one :(

**Comment by HowardWolosky** [comment]
> Ran the whole test suite for this PR, and the GitHubRepositoryForks test has deleted my fork of the PowerShellForGitHub repo, which includes all the branches for the pending PRs here including this one :(

Oh no.  I've immediately merged in #251 to avoid this in the future.  So sorry about that.

**Comment by X-Guardian** [comment]
I've still got the branches locally, so no lost work, and I've pushed them back to my new fork of the repo, but the links to the open PRs are broken. PRs affected are:

- #200 - Branch: https://github.com/X-Guardian/PowerShellForGitHub/tree/GitHubBranches-NewRemove
- #222 - Branch: https://github.com/X-Guardian/PowerShellForGitHub/tree/GitHubTeams-Add-New-Remove
- #238 - Branch: https://github.com/X-Guardian/PowerShellForGitHub/tree/GitHubRepositoryBranchProtection
- #248 - Branch: https://github.com/X-Guardian/PowerShellForGitHub/tree/ShouldProcess-Update

Any idea how to fix the PRs?

**Comment by HowardWolosky** [comment]
> Any idea how to fix the PRs?

Not sure yet.  Doing some searching online.  Will be able to look a bit more this afternoon.

**Comment by HowardWolosky** [comment]
I don't think you can.

https://stackoverflow.com/questions/36071272/fix-unknown-repository-of-an-opened-pr-after-deleted-the-fork
https://github.com/isaacs/github/issues/168

I think you'll just need to create new pull requests for those outstanding ones, and then close out the ones that have now been orphaned.

File: GitHubProjectColumns.ps1 [hunk 1]
```diff
@@ -473,6 +477,13 @@ filter Move-GitHubProjectColumn
         'position' = $Position
     }
 
+    if (-not $PSCmdlet.ShouldProcess($ColumnName, 'Move GitHub Project Column'))
```
Comment (HowardWolosky) [inline GitHubProjectColumns.ps1:480]:
```suggestion
    if (-not $PSCmdlet.ShouldProcess($Column, 'Move GitHub Project Column'))
```

**Comment by HowardWolosky** [comment]
Actually, looking at this a bit more as full files as opposed to just the diffs, I think the one thing about the pattern that I'd change is that I'd move the check to the top of the functions as opposed to right before the call to `Invoke-GHRestMethod*`.

File: GitHubPullRequests.ps1 [hunk 2]
```diff
@@ -422,6 +416,13 @@ filter New-GitHubPullRequest
         $acceptHeader = 'application/vnd.github.shadow-cat-preview+json'
     }
 
+    if (-not $PSCmdlet.ShouldProcess($Title, 'Create GitHub Pull Request'))
```
Comment (HowardWolosky) [inline GitHubPullRequests.ps1:419]: Not guaranteed to have a `Title`.  The user might be creating a pull request from an `Issue`.

File: GitHubIssueComments.ps1 [hunk 3]
```diff
@@ -552,6 +550,13 @@ filter Set-GitHubIssueComment
         'body' = $Body
     }
 
+    if (-not $PSCmdlet.ShouldProcess($Issue, 'Create GitHub Issue Comment'))
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubIssueComments.ps1:553]: Done

File: GitHubProjectColumns.ps1 (hunk 1 above)
Comment (X-Guardian) [reply GitHubProjectColumns.ps1:480]: Done

File: GitHubPullRequests.ps1 (hunk 2 above)
Comment (X-Guardian) [reply GitHubPullRequests.ps1:419]: Code refactored.

**Comment by X-Guardian** [comment]
> Actually, looking at this a bit more as full files as opposed to just the diffs, I think the one thing about the pattern that I'd change is that I'd move the check to the top of the functions as opposed to right before the call to `Invoke-GHRestMethod*`.

If you are running a function with the `WhatIf` parameter, the main purpose is to validate that the set of parameters you have passed are valid. To achieve this, the `$PSCmdlet.ShouldProcess` statement needs to occur after all the parameter validation code.

File: GitHubConfiguration.ps1 [hunk 4]
```diff
@@ -246,6 +244,11 @@ function Set-GitHubConfiguration
         }
     }
 
+    if (-not $PSCmdlet.ShouldProcess('GitHubConfiguration',' Save'))
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubConfiguration.ps1:247]: Done

File: GitHubConfiguration.ps1 [hunk 5]
```diff
@@ -342,6 +344,11 @@ function Save-GitHubConfiguration
         [string] $Path
     )
 
+    if (-not $PSCmdlet.ShouldProcess('GitHub Configuration','Save'))
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubConfiguration.ps1:347]: Done

File: GitHubConfiguration.ps1 [hunk 6]
```diff
@@ -507,12 +514,14 @@ function Reset-GitHubConfiguration
 
     Set-TelemetryEvent -EventName Reset-GitHubConfiguration
 
+    if (-not $PSCmdlet.ShouldProcess('GitHub Configuration','Reset'))
```
[the remark this answers is not in the document]
Comment (X-Guardian) [reply GitHubConfiguration.ps1:517]: Done

**Comment by HowardWolosky** [comment]
> If you are running a function with the `WhatIf` parameter, the main purpose is to validate that the set of parameters you have passed are valid. To achieve this, the `$PSCmdlet.ShouldProcess` statement needs to occur after all the parameter validation code.

Sufficiently compelling argument for me.  Please keep as-is.

**Comment by X-Guardian** [comment]
This PR is superseded by PR #254

**Pull request closed without merging**